kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/Metadata.kt

```diff
@@ -0,0 +1,350 @@
+import kotlinx.metadata.KmValueParameterVisitor
+import kotlinx.metadata.KmVariance
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import javax.lang.model.element.Element
+import javax.lang.model.element.ExecutableElement
+
+@Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
+internal fun Element.readMetadata(): KotlinClassHeader? {
+  return getAnnotation(Metadata::class.java)?.asClassHeader()
+}
+
+@Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
+internal fun Class<*>.readMetadata(): KotlinClassHeader? {
+  return getAnnotation(Metadata::class.java)?.asClassHeader()
+}
+
+@Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
+internal fun Metadata.asClassHeader(): KotlinClassHeader {
+  return KotlinClassHeader(k, mv, bv, d1, d2, xs, pn, xi)
+}
+
+internal fun KotlinClassHeader.readKotlinClassMetadata(): KotlinClassMetadata? {
+  return KotlinClassMetadata.read(this)
+}
+
+internal fun KmVariance.asKModifier(): KModifier? {
+  return when (this) {
+    KmVariance.IN -> KModifier.IN
+    KmVariance.OUT -> KModifier.OUT
+    KmVariance.INVARIANT -> null
+  }
+}
+
+/**
+ * Resolves the TypeName of this type as it would be seen in the source code, including nullability
+ * and generic type parameters.
+ *
+ * @param flags the [Flags] associated with this type
+ * @param [getTypeParameter] a function that returns the type parameter for the given index. **Only
+ *     called if [TypeNameKmTypeVisitor.visitTypeParameter] is called**
+ * @param useTypeAlias indicates whether or not to use type aliases or resolve their underlying
+ *     types
+ */
+internal class TypeNameKmTypeVisitor(flags: Flags,
+    private val getTypeParameter: ((index: Int) -> TypeName)? = null,
+    private val useTypeAlias: Boolean = false,
+    private val receiver: (TypeName) -> Unit) : KmTypeVisitor() {
+
+  private val nullable = Flag.Type.IS_NULLABLE(flags)
+  private var className: String? = null
+  private var typeAliasName: String? = null
+  private var typeAliasType: TypeName? = null
+  private var flexibleTypeUpperBound: TypeName? = null
+  private var outerType: TypeName? = null
+  private var typeParameter: TypeName? = null
+  private val argumentList = mutableListOf<TypeName>()
+
+  override fun visitAbbreviatedType(flags: Flags): KmTypeVisitor? {
+    if (!useTypeAlias) {
+      return null
+    }
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      typeAliasType = it
+    }
+  }
+
+  override fun visitArgument(flags: Flags, variance: KmVariance): KmTypeVisitor? {
+    return TypeNameKmTypeVisitor(flags, getTypeParameter, useTypeAlias) {
+      argumentList.add(
+          when (variance) {
+            KmVariance.IN -> WildcardTypeName.supertypeOf(it)
+            KmVariance.OUT -> {
+              if (it == ANY) {
+                // This becomes a *, which we actually don't want here.
+                // List<Any> works with List<*>, but List<*> doesn't work with List<Any>
```

Comment:
possibly. Writing code this way is something the IDE will yell at you for being redundant, but is technically legal.

From the smoke test type:

```kotlin
val any: List<*>,
val anyTwo: List<Any>,
val anyOut: List<out Any>,
```

in the generated adapter is handled as 

```kotlin
private val listOfAnyAdapter: JsonAdapter<List<*>> =
        moshi.adapter<List<*>>(Types.newParameterizedType(List::class.java, Types.subtypeOf(Any::class.java))).nonNull()

private val listOfAnyAdapter_: JsonAdapter<List<Any>> =
        moshi.adapter<List<Any>>(Types.newParameterizedType(List::class.java, Any::class.java)).nonNull()
```

Where `any` uses `listOfAnyAdapter`, and both `anyTwo` and `anyOut` use `listOfAnyAdapter_`. Kotlin sees `List<out Any>` and `List<Any>` as the same, and different than `List<*>`. The distinction seems to be that `*` to the compiler is actually equivalent to `Types.subtypeOf(Any::class.java)`, but KotlinPoet takes `Any` and converts it to a *.

So... maybe this is something we should fix in KotlinPoet, but it's super special cased. Basically it would say that `WildcardTypeName.subtypeOf(it)` should be `*` IFF `it` isn't `Any` with `out` projection.